cajita/src/Cajita_SparseIndexSpace.hpp

```diff
@@ -0,0 +1,890 @@
+// Naming:
+//      Block = MPI Rank
+//      Tile = Sub-block with several cells
+//      Cell = Basic grid unit
+//---------------------------------------------------------------------------//
+
+//! Declaration of BlockIndexSpace
+template <typename MemorySpace, unsigned long long CBits,
+          unsigned long long CNumPerDim, unsigned long long CNumPerTile,
+          HashTypes Hash, typename Key, typename Value>
+class BlockIndexSpace;
+
+//! Declaration of TileIndexSpace
+template <int CBits, int CNumPerDim, int CNumPerTile>
+class TileIndexSpace;
+
+/*!
+  \class SparseIndexSpace
+  \brief Sparse index space, with a hierarchical structure (cell->tile->block)
+  \tparam MemorySpace Memory space to store the IndexSpace(Hash Table)
+  \tparam CellPerTileDim Cell number inside each tile per dimension
+  \tparam Hash Hash type (lexicographical or morton)
+  \tparam Key Type of the tile/cell hash key
+  \tparam Value Type of the tile/cell No.
+ */
+template <typename MemorySpace, unsigned long long CellPerTileDim = 4,
+          HashTypes Hash = HashTypes::Naive, typename Key = uint64_t,
+          typename Value = uint64_t>
+class SparseIndexSpace
+{
+  public:
+    //! Number of dimensions, 3 = ijk
+    static constexpr int rank = 3;
+    //! Number of bits (per dimension) needed to index the cells inside a tile
+    static constexpr unsigned long long cell_bits_per_tile_dim =
+        bitCount( CellPerTileDim );
+    //! Number of cells inside each tile (per dimension), tile size reset to
+    //! power of 2
+    static constexpr unsigned long long cell_num_per_tile_dim =
+        1 << cell_bits_per_tile_dim;
+    //! Cell mask (per dimension), indicating which part of the tile
+    //! address(binary bits) will index the cells inside each tile
+    static constexpr unsigned long long cell_mask_per_tile_dim =
+        ( 1 << cell_bits_per_tile_dim ) - 1;
+    //! Number of bits (total) needed to index the cells inside a tile
+    static constexpr unsigned long long cell_bits_per_tile =
+        cell_bits_per_tile_dim + cell_bits_per_tile_dim +
+        cell_bits_per_tile_dim;
+    //! Number of cells (total) inside each tile
+    static constexpr unsigned long long cell_num_per_tile =
+        cell_num_per_tile_dim * cell_num_per_tile_dim * cell_num_per_tile_dim;
+    //! Types
+    using key_type = Key;                        // tile hash key
+    using value_type = Value;                    // tile No.
+    static constexpr HashTypes hash_type = Hash; // hash table
+
+    /*!
+      \brief (Host) Constructor
+      \param size The size of the block (MPI rank) (Unit: cell)
+      \param capacity Expected capacity of the allocator to store the tiles
+      \param rehash_factor The factor multiplies to the capacity for resizing,
+      when tile nums exceed the capacity
+    */
+    SparseIndexSpace( const std::array<int, rank> size,
+                      const unsigned int capacity, const float rehash_factor )
```

Comment:
Can we just eliminate `rehash_factor`? It isn't used in your implementation as far as I can tell.

Reply:
Sure!